### PR TITLE
[Python] data subscription pages batch [Replace C# samples]

### PR DESCRIPTION
**Pull request by reebhub**
[RDoc-2925](https://issues.hibernatingrhinos.com/issue/RDoc-2925/Python-Client-API-Data-subscriptions-Creation-how-to-create-data-subscription-Replace-C-samples) Data subscriptions > Creation > How to create data subscription
[RDoc-2917](https://issues.hibernatingrhinos.com/issue/RDoc-2917/Python-Client-API-Data-subscriptions-Creation-Examples-Replace-C-samples) Data subscriptions > Creation > Examples
[RDoc-2918](https://issues.hibernatingrhinos.com/issue/RDoc-2918/Python-Client-API-Data-subscriptions-Creation-API-overview-Replace-C-samples) Data subscriptions > Creation > API overview
[RDoc-2919](https://issues.hibernatingrhinos.com/issue/RDoc-2919/Python-Client-API-Data-subscriptions-Consumption-How-to-consume-Replace-C-samples) Data subscriptions > Consumption > How to consume
[RDoc-2920](https://issues.hibernatingrhinos.com/issue/RDoc-2920/Python-Client-API-Data-subscriptions-Consumption-Examples-Replace-C-samples) Data subscriptions > Consumption > Examples
[RDoc-2921](https://issues.hibernatingrhinos.com/issue/RDoc-2921/Python-Client-API-Data-subscriptions-Consumption-API-overview-Replace-C-samples) Data subscriptions > Consumption > API overview

### CODE REVIEW

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+{PANEL/}
+
+{PANEL:`SubscriptionWorkerOptions`}
+
+{NOTE The only mandatory parameter for the ceation of `SubscriptionWorkerOptions` is the subscription **name**. /}
+
+| Member | Type | Description |
+|--------|:-----|-------------| 
+| **subscription_name** | `str` | Returns the subscription name passed to the constructor. This name will be used by the server side to identify the subscription. |
+| **time_to_wait_before_connection_retry** | `DateTime` | Time to wait before reconnecting, in the case of non-aborting failure during the subscription processing.<br>Default: 5 seconds. |
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:47]:
```suggestion
| **time_to_wait_before_connection_retry** | `datetime` | Time to wait before reconnecting, in the case of non-aborting failure during the subscription processing.<br>Default: 5 seconds. |

```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+| **receive_buffer_size_in_bytes** | `int` | Size, in bytes, of the TCP socket buffer used for _receiving_ data.<br>Default: 32,768 (32 KiB) |
+
+{PANEL/}
+
+{PANEL:Running subscription worker}
+
+After [generating](../../../client-api/data-subscriptions/consumption/api-overview#subscription-worker-generation) 
+a subscription worker, the worker doesn't process any documents yet.  
+Processing starts when the `run` function is called.  
+The `run` function receives the client-side code as a delegate that will process the retrieved batches:
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:62]:
```suggestion
The `run` function receives the client-side code as a function that will process the retrieved batches:

```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+As long as there is no exception, the worker will continue addressing the same 
+server that the first batch was received from.  
+If the worker fails to reach that node, it will try to 
+[failover](../../../client-api/configuration/load-balance/overview) to another node 
+from the session's topology list.  
+The node that the worker succeeds connecting to, will inform the worker which 
+node is currently responsible for data subscriptions.  
+{NOTE/}
+
+| `SubscriptionBatch[_T].item` Member | Type | Description |
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:89]:
```suggestion
| `SubscriptionBatch[_T].Item` Member | Type | Description |

```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+{PANEL:`SubscriptionWorker[_T]`}
+
+---
+
+### Methods:
+
+| Method | Return Type | Description |
+|--------|:-----|-------------| 
+| `dispose` | `void` | Aborts subscription worker operation ungracefully by waiting for the task returned by the `run` function to finish running. |
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:114]:
```suggestion
| `close` | `void` | Aborts subscription worker operation ungracefully by waiting for the task returned by the `run` function to finish running. |

```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+| `dispose(bool waitForSubscriptionTask)` | `void` | Aborts the subscription worker, but allows deciding whether to wait for the `run` function task or not. |
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:115]:
```suggestion

```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+| `run` | `Task` | Starts the subscription worker work of processing batches, receiving the batch processing delegates (see [above](../../../client-api/data-subscriptions/consumption/api-overview#running-subscription-worker)). |
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:116]:
```suggestion
| `run` | `Future[None]` | Starts the subscription worker work of processing batches, receiving the batch processing delegates (see [above](../../../client-api/data-subscriptions/consumption/api-overview#running-subscription-worker)). |

```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+---
+
+### Events:
+
+| Event | Type\Return type | Description |
+|--------|:-----|-------------| 
+| **after\_acknowledgment** | `AfterAcknowledgmentAction` (event) | Event invoked after each time the server acknowledges batch processing progress. |
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:124]:
```suggestion
| **after\_acknowledgment** | `Callable[[SubscriptionBatch[_T]], None]` | Event invoked after each time the server acknowledges batch processing progress. |

```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+| `AfterAcknowledgmentAction` Parameters | | |
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:126]: maybe
```suggestion
| `Callable[[SubscriptionBatch[_T]], None]` Parameters | | |

```

but better 
```suggestion
| `after_acknowledgment` Parameters | | |

```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+| ------------- | ------------- | ----- |
+| **batch** | `SubscriptionBatch[_T]` | The batch process which was acknowledged |
+
+| Return value | |
+| ------------- | ----- |
+| `Task` | The worker waits for the task to finish the event processing |
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:132]:
```suggestion
| `Future[None]` | The worker waits for the task to finish the event processing |

```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+If an exception is thrown, the worker will abort the current batch process. 
+A worker can be configured to treat the thrown exception in either of the following two ways:  
+
+* By default, the worker will wrap the thrown exception with a `SubscriberErrorException` exception 
+  and rethrow it, terminating the subscription execution without acknowledging progress or retrying.  
+  The task returned by the `run` function will be terminated with an erroneous state, throwing 
+  a `SubscriberErrorException` exception.  
+
+* If `SubscriptionWorkerOptions`'s value `IgnoreSubscriberErrors` is set to True, the erroneous
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown:68]:
```suggestion
* If `SubscriptionWorkerOptions`'s value `ignore_subscriber_errors` is set to True, the erroneous 
```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+
+* If `SubscriptionWorkerOptions`'s value `IgnoreSubscriberErrors` is set to True, the erroneous 
+  batch will get acknowledged without retrying and the next batches will continue processing. 
+{INFO/}
+
+{INFO: Reconnecting}
+The above cases describe situations in which a worker tries to reconnect with the server.  
+There are two key `SubscriptionWorkerOptions` fields controlling this state:
+
+* `TimeToWaitBeforeConnectionRetry` - Worker 'sleep' period before trying to reconnect.
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown:76]:
```suggestion
* `time_to_wait_before_connection_retry` - Worker 'sleep' period before trying to reconnect.  
```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+* If `SubscriptionWorkerOptions`'s value `IgnoreSubscriberErrors` is set to True, the erroneous 
+  batch will get acknowledged without retrying and the next batches will continue processing. 
+{INFO/}
+
+{INFO: Reconnecting}
+The above cases describe situations in which a worker tries to reconnect with the server.  
+There are two key `SubscriptionWorkerOptions` fields controlling this state:
+
+* `TimeToWaitBeforeConnectionRetry` - Worker 'sleep' period before trying to reconnect.  
+* `MaxErroneousPeriod` - Maximum time that the worker is allowed to be in an erroneous state.
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown:77]:
```suggestion
* `max_erroneous_period` - Maximum time that the worker is allowed to be in an erroneous state.  
```

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+The above cases describe situations in which a worker tries to reconnect with the server.  
+There are two key `SubscriptionWorkerOptions` fields controlling this state:
+
+* `TimeToWaitBeforeConnectionRetry` - Worker 'sleep' period before trying to reconnect.  
+* `MaxErroneousPeriod` - Maximum time that the worker is allowed to be in an erroneous state.  
+  When this time elapses, the worker will stop trying to reconnect.  
+{INFO/}
+
+{INFO: OnUnexpectedSubscriptionError}
+`OnUnexpectedSubscriptionError` is the event raised when a connection failure occurs
```
Comment (poissoncorp) [inline Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown:82]:
```suggestion
`on_unexpected_subscription_error` is the event raised when a connection failure occurs 
```

**Review by poissoncorp — CHANGES_REQUESTED** [review]
let's address the comments & we're good to merge

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown (hunk 1 above)
Comment (reebhub) [reply Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:47]: done

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown (hunk 2 above)
Comment (reebhub) [reply Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:62]: done

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown (hunk 3 above)
Comment (reebhub) [reply Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:89]: done

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown (hunk 4 above)
Comment (reebhub) [reply Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:114]: done

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown (hunk 6 above)
Comment (reebhub) [reply Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:116]: done

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown (hunk 8 above)
Comment (reebhub) [reply Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:126]: done (the second one :o)

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown (hunk 9 above)
Comment (reebhub) [reply Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/api-overview.python.markdown:132]: done

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown (hunk 10 above)
Comment (reebhub) [reply Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown:68]: done

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown (hunk 11 above)
Comment (reebhub) [reply Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown:76]: done

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown (hunk 12 above)
Comment (reebhub) [reply Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown:77]: done

File: Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown (hunk 13 above)
Comment (reebhub) [reply Documentation/5.4/Raven.Documentation.Pages/client-api/data-subscriptions/consumption/how-to-consume-data-subscription.python.markdown:82]: done

**Pull request merged**